WIP: TranscodeSub

j0sh opened the pull request:
Implements https://github.com/livepeer/go-livepeer-basicnet/issues/29

Notes:
* While the  external API is a `TranscodeSub` method on VideoNetwork. This comes rather late in the commit series. I should squash the earlier commits for a cleaner history. https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/242fee8cd3601bdddb87888a1cb553814fcedfb5
* We reuse the libp2p listener that the transcoder already has open for the direct connection
* Transcoders handle the direct connection via implementing Notifiee on its own subscriber. https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/8da37d4f63899abd887e2ef2e511dd336f9c3c7e
* Updates libp2p in order to work around an issue with the peerstore that was affecting tests.

j0sh committed:
WIP: Basic TranscodeSub.

j0sh committed:
test: Add tests for TranscodeSub gob codecs.

j0sh committed:
WIP: Direct connections after TranscodeSub.

j0sh committed:
Subscribe transcoder after direct connection from broadcaster.

j0sh committed:
Skip TranscodeSub if a direct peer already exists.

Subscribe directly instead.

ericxtang commented on basic_subscriber.go:
```diff
@@ -79,10 +81,43 @@ func (s *BasicSubscriber) Subscribe(ctx context.Context, gotData func(seqNo uint
+					return err
+				}
+				maddrs := make([]ma.Multiaddr, len(s.Network.NetworkNode.Host().Addrs()))
+				for i, v := range s.Network.NetworkNode.Host().Addrs() {
```
I haven't used encapsulation before.  Is there a reason for using the IPFS encapsulation instead of sending the nodeID and multiaddr both in `TranscodeSub`?

j0sh replied:
No reason other than an overly literal reading of what we discussed in https://github.com/livepeer/go-livepeer-basicnet/issues/21. Fixed to add a `NodeID` to the struct rather than appending it to each multiaddr in the list. Good call.

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/70f8fe7e4f283e35a5b6ce8f878f9efcc2a6c735

ericxtang commented on basic_subscriber.go:
```diff
@@ -79,10 +81,43 @@ func (s *BasicSubscriber) Subscribe(ctx context.Context, gotData func(seqNo uint
 		glog.V(5).Infof("New peer from kademlia: %v", peer.IDHexEncode(p))
 		ns := s.Network.NetworkNode.GetOutStream(p)
 		if ns != nil {
```
Is `Subscribe` the right place to send `TranscodeSub`?  Nodes that are not transcoders can also try and subscribe to the video, and I don't think they should be able to get a direct connection with the broadcaster.

What do you think about adding `TranscodeSub(StrmID)` to the `VideoNetwork` interface, so the node will only call it when it's a transcoder?  

I also think this message should be an enhancement rather than a requirement - meaning if for some reason the broadcaster cannot create a direct connection with the transcoder, everything should still work in the relayer-based model.

j0sh replied:
You're right that is probably a better approach. This setup was based on a brief discussion we had in gitter, which I likely mis-interpreted.

Agreed on the relay based fallback; that is on the to-do list ('Resilience' bullet point in https://github.com/livepeer/go-livepeer-basicnet/issues/29). In order to do that, the broadcaster needs to nack the TranscodeSub request, so we need an additional message.

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/242fee8cd3601bdddb87888a1cb553814fcedfb5

ericxtang reviewed (commented):
Let some comments

j0sh committed:
Replace IPFS multiaddr with NodeID.

j0sh committed:
Move TranscodeSub into its own public API.

j0sh committed:
test: Broadcaster-transcoder subs initially connected nodes.

j0sh committed:
test: Broadcaster-transcoder subs on disconnected nodes.

j0sh committed:
Make TranscodeSub behave the same as SubReq.

Avoids the need for another round-trip and avoids the nack problem.

j0sh committed:
test: More sanity checks for TranscodeSub.

j0sh committed:
test: Relay fallback for unreachable transcoders.

j0sh committed:
gx: Update package.json with newer versions.

Fixes an error during the test suite that resulted from a bug
within go-peerstore.

j0sh committed:
gx: Update deps.

j0sh committed:
dht: Fix record validator.

The record validation API changed here:
https://github.com/libp2p/go-libp2p-record/commit/1c075e0108b178f30857ca532ecf1202388cbde6

j0sh committed:
README: gx update 'documentation'.

j0sh committed:
Cancel relay sub upon direct connection.

j0sh committed:
test: Check for relay sub cancellation.

j0sh committed:
TranscodeSub: Periodic retransmission.

Needed if a broadcaster misses the first few transmissions.

j0sh committed:
test: Fix up testcases that failed after TranscodeSub went async.

j0sh committed:
test: Add cases for TranscodeSub retransmits.

j0sh committed:
test: Ignore broken testcases.

ericxtang commented on network_node.go:
```diff
@@ -47,6 +52,8 @@ type BasicNetworkNode struct {
 	outStreams     map[peer.ID]*BasicOutStream
 	outStreamsLock *sync.Mutex
+	sign           func(data []byte) ([]byte, error)
+	verify         func(data []byte, sig []byte, strmID string) bool
```
Can we set a default "verify" here that just returns true?  This way the default behavior is `pass`.

j0sh replied:
Generally I don't like enabling insecure defaults, but in this case we don't have a choice until the Livepeer client can actually do verification. If we can remove the default behavior after client support is in, that'd be OK.

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/bdb52f96dab751294da9bce3fd2a5ba0751dcbfc

ericxtang commented on basic_network.go:
```diff
@@ -782,6 +805,51 @@ func handleTranscodeResponse(nw *BasicVideoNetwork, remotePID peer.ID, tr Transc
+func handleTranscodeSub(nw *BasicVideoNetwork, conn net.Conn, ts TranscodeSubMsg) error {
+	glog.Infof("%v In handleTranscodeSub; received from %v", nw.NetworkNode.ID(), conn.RemotePeer())
+	ok := nw.NetworkNode.VerifyTranscoderSig(ts.BytesForSigning(), ts.Sig, ts.StrmID)
+	if !ok {
```
Anything stopping us from returning the error here?

j0sh replied:
Do you mean returning an error from the VerifyTranscoderSig function [1], or the error that's commented out a few lines down [2]?

[1] Is this more idiomatic Go? Normally that'd probably be a better idea. The only incongruity I can think of is related to the comment that the message handler expects `ErrHandleMsg` ; we wouldn't be able to control the type of error returned. We'd need to take the error, discard it, then return another, which loses the richer context that error propagation offers, and 'correct' usage becomes murkier. With a bool though, there's no question about the semantics (which matter here for errors).

[2] With a default verify function, sure.

j0sh replied:
Fixed [2] in https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/bdb52f96dab751294da9bce3fd2a5ba0751dcbfc

ericxtang commented on basic_network.go:
```diff
@@ -782,6 +805,51 @@ func handleTranscodeResponse(nw *BasicVideoNetwork, remotePID peer.ID, tr Transc
+	mas := make([]string, len(ts.MultiAddrs))
+	if len(ts.MultiAddrs) <= 0 {
+		glog.Errorf("No multiaddrs; can't establish direct cxn to transcoder")
+		return ErrBadMultiaddrs
```
Currently we reset the stream (similar to dropping the connection) if we return anything other than ErrHandleMsg in the handlers - not sure we should do that here. Reference: `SetupProtocol()`

j0sh replied:
Fixed in https://github.com/livepeer/go-livepeer-basicnet/pull/32/commits/bdb52f96dab751294da9bce3fd2a5ba0751dcbfc

ericxtang commented on basic_subscriber.go:
```diff
@@ -46,6 +55,48 @@ func (s *BasicSubscriber) InsertData(sd *StreamDataMsg) error {
+	s.Network.NetworkNode.Host().Network().Notify(s)
+	glog.Infof("%v Sending TranscodeSub message for %v", s.Network.NetworkNode.ID(), s.StrmID)
+
+	// periodic retransmit in case the broadcaster misses the first message(s)
```
Is re-transmit the proper strategy here?  What would make a broadcaster miss the message the first time, and get it in the future?  In the current implementation, when will sendCtx.Done() be triggered?

ps - I've been using https://github.com/cenkalti/backoff for a lot of the retry stuff for Infura websockets, might save you some cruft code.

j0sh replied:
> What would make a broadcaster miss the message the first time, and get it in the future?

With the current peer routing mechanism, we have no feedback on delivery or message propagation, and no visibility into the network's connectivity. There could be intermittent connectivity anywhere along the route, and retrying gives the topology time to heal and work around routing partitions (especially since we only forward to one peer at a time).

While hopefully this issue wont be too common, we do want to be prepared when it happens; otherwise it'd be extremely difficult to diagnose, and frustrating for a user who just spent gas on a job transaction.

> Is re-transmit the proper strategy here?

What are the other options?

> will sendCtx.Done() be triggered?

Three possible points/scenarios where sendCtx.Done would be triggered:

* After the direct connection is established. This is an indicator that everything went normally.
* After receiving a data message. If no direct connection was received, this usually means the transcoder is unreachable from the broadcaster.
* If the subscribe context gets cancelled for whatever reason, before the first two cases above.

>  I've been using https://github.com/cenkalti/backoff for a lot of the retry stuff for Infura websockets

Looked into it briefly, but I'm not keen on pulling an entire dependency just to save a few lines. Maybe if we have to re-use this logic later.

ericxtang reviewed (changes requested):
Left some comments

j0sh committed:
Pass transocder sig verification by default.

j0sh committed:
test: Add case for transcoder sig verification.

j0sh committed:
test: Split transcoder subs into their own file.

Also add some logging prior to tests.

j0sh committed:
Clean up unused relayers.

j0sh committed:
test: Accommodate correctly cleaned relayers.

j0sh committed:
One last bit of unused code cleanup.

j0sh committed:
Run test files individually on CircleCI.

Exquisitely fragile; success depends on local invocation order.

Also re-enable the disabled tests, because now it Works Here.
